File: src/main/java/com/trunkshell/testzilla/model/Product.java
package com.trunkshell.testzilla.model;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Table;

import org.hibernate.annotations.Formula;

import com.fasterxml.jackson.annotation.JsonIgnore;

/**
 * 产品的Model.
 * @author Xie Haozhe
 */
@Entity
@Table(name = "tz_products")
public class Product implements Serializable {
	/**
	 * Product的默认构造函数.
	 */
	public Product() {}
	
	/**
	 * Product的构造函数.
	 * @param productName - 产品的名称
	 * @param productLogo - 产品Logo的相对路径
	 * @param productCategory - 产品的分类
	 * @param latestVersion - 产品最新版本号
	 * @param developer - 产品的开发者
	 * @param prerequisites - 产品使用的先决条件
	 * @param url - 产品安装包的相对路径或网站URL
	 * @param description - 产品的相关描述
	 */
	public Product(String productName, String productLogo, ProductCategory productCategory, 
				   String latestVersion, User developer, String prerequisites, String url, 
				   String description) {
		this.productName = productName;
		this.productLogo = productLogo;
		this.productCategory = productCategory;
		this.latestVersion = latestVersion;
		this.developer = developer;
		this.prerequisites = prerequisites;
		this.url = url;
		this.description = description;
	}
	
	/**
	 * Product的构造函数(用于单元测试).
	 * @param productId - 产品的唯一标识符
	 * @param productName - 产品的名称
	 * @param productLogo - 产品Logo的相对路径
	 * @param productCategory - 产品的分类
	 * @param latestVersion - 产品最新版本号
	 * @param developer - 产品的开发者
	 * @param prerequisites - 产品使用的先决条件
	 * @param url - 产品安装包的相对路径或网站URL
	 * @param description - 产品的相关描述
	 */
	public Product(long productId, String productName, String productLogo, ProductCategory productCategory, 
			   String latestVersion, User developer, String prerequisites, String url, 
			   String description) {
		this(productName, productLogo, productCategory, latestVersion, developer, prerequisites, url, description);
		this.productId = productId;
	}

	/**
	 * 获取产品的唯一标识符.
	 * @return 产品的唯一标识符
	 */
	public long getProductId() {
		return productId;
	}

	/**
	 * 设置产品的唯一标识符.
	 * @param productId - 产品的唯一标识符
	 */
	public void setProductId(long productId) {
		this.productId = productId;
	}

	/**
	 * 获取产品的名称.
	 * @return 产品的名称
	 */
	public String getProductName() {
		return productName;
	}

	/**
	 * 设置产品的名称.
	 * @param productName - 产品的名称
	 */
	public void setProductName(String productName) {
		this.productName = productName;
	}

	/**
	 * 获取产品Logo的相对路径.
	 * @return 产品Logo的相对路径
	 */
	public String getProductLogo() {
		return productLogo;
	}

	/**
	 * 设置产品Logo的相对路径.
	 * @param productLogo - 产品Logo的相对路径
	 */
	public void setProductLogo(String productLogo) {
		this.productLogo = productLogo;
	}

	/**
	 * 获取产品的分类.
	 * @return 产品的分类
	 */
	public ProductCategory getProductCategory() {
		return productCategory;
	}

	/**
	 * 设置产品的分类.
	 * @param productCategory - 产品的分类
	 */
	public void setProductCategory(ProductCategory productCategory) {
		this.productCategory = productCategory;
	}
	
	/**
	 * 获取产品最新版本号.
	 * @return 产品最新版本号
	 */
	public String getLatestVersion() {
		return latestVersion;
	}

	/**
	 * 设置产品最新版本号.
	 * @param productLatestVersion - 产品最新版本号
	 */
	public void setLatestVersion(String latestVersion) {
		this.latestVersion = latestVersion;
	}

	/**
	 * 获取产品开发者.
	 * @return 产品开发者
	 */
	@JsonIgnore
	public User getDeveloper() {
		return developer;
	}

	/**
	 * 设置产品开发者.
	 * @param developer - 产品开发者
	 */
	public void setDeveloper(User developer) {
		this.developer = developer;
	}

	/**
	 * 获取产品使用的先决条件.
	 * @return 产品使用的先决条件
	 */
	public String getPrerequisites() {
		return prerequisites;
	}

	/**
	 * 设置产品使用的先决条件.
	 * @param prerequisites - 产品使用的先决条件
	 */
	public void setPrerequisites(String prerequisites) {
		this.prerequisites = prerequisites;
	}

	/**
	 * 获取产品安装包的相对路径或网站URL.
	 * @return 产品安装包的相对路径或网站URL
	 */
	public String getUrl() {
		return url;
	}

	/**
	 * 设置产品安装包的相对路径或网站URL.
	 * @param url - 产品安装包的相对路径或网站URL
	 */
	public void setUrl(String url) {
		this.url = url;
	}

	/**
	 * 获取产品的相关描述.
	 * @return 产品的相关描述
	 */
	public String getDescription() {
		return description;
	}

	/**@JsonInclude(JsonInclude.Include.NON_NULL)
	 * 设置产品的相关描述.
	 * @param description - 产品的相关描述
	 */
	public void setDescription(String description) {
		this.description = description;
	}

	/**
	 * 获取Bug列表(以便1-N关联).
	 * @return Bug列表
	 */
	@JsonIgnore
	public List<Bug> getBugs() {
		return bugs;
	}

	/**
	 * 设置Bug列表.
	 * @param bugs - Bug列表
	 */
	public void setBugs(List<Bug> bugs) {
		this.bugs = bugs;
	}

	/**
	 * 获取已发现的Bug数量.
	 * @return 已发现的Bug数量
	 */
	public int getNumberOfIssues() {
		return numberOfIssues;
	}
	
	/* (non-Javadoc)
	 * @see java.lang.Object#equals(java.lang.Object)
	 */
	@Override
	public boolean equals(Object o) {
		if ( !(o instanceof Product) ) {
			return false;
		}
		return this.productId == ((Product)o).productId;
	}

	/* (non-Javadoc)
	 * @see java.lang.Object#toString()
	 */
	public String toString() {
		return String.format("Product: [Id=%s, Name=%s, productCategory={%s}, latestVersion=%s, developer={%s}, prerequisites=%s, url=%s]", 
				new Object[] { productId, productName, productCategory, latestVersion, developer, prerequisites, url });
	}

	/**
	 * 产品的唯一标识符.
	 */
	@Id
	@GeneratedValue
	@Column(name = "product_id")
	private long productId;
	
	/**
	 * 产品的名称.
	 */
	@Column(name = "product_name")
	private String productName;
	
	/**
	 * 产品Logo的相对路径.
	 */
	@Column(name = "product_logo")
	private String productLogo;
	
	/**
	 * 产品的分类.
	 */
	@ManyToOne(targetEntity = ProductCategory.class)
	@JoinColumn(name = "product_category_id")
	private ProductCategory productCategory;
	
	/**
	 * 产品最新版本号.
	 */
	@Column(name = "product_latest_version")
	private String latestVersion;
	
	/**
	 * 产品的开发者.
	 */
	@ManyToOne(targetEntity = User.class)
	@JoinColumn(name = "product_developer_id")
	private User developer;
	
	/**
	 * 产品使用的先决条件.
	 */
	@Column(name = "product_prerequisites")
	private String prerequisites;
	
	/**
	 * 产品安装包的相对路径或网站URL.
	 */
	@Column(name = "product_url")
	private String url;
	
	/**
	 * 产品的相关描述.
	 */
	@Column(name = "product_description")
	private String description;
	
	/**
	 * Bug列表(以便1-N关联).
	 */
	@OneToMany(targetEntity = Bug.class, 
				fetch = FetchType.LAZY, mappedBy = "product")
	private List<Bug> bugs = new ArrayList<Bug>();
	
	/**
	 * 已发现的Bug数量.
	 */
	@Formula(value = "(SELECT COUNT(*) FROM tz_bugs b WHERE b.product_id = product_id)")
	private int numberOfIssues;
	
	/**
	 * 唯一的序列化标识符.
	 */
	private static final long serialVersionUID = 2604441212604210469L;
}

File: app/models/Product.php
<?php
/**
 * @Author: Haozhe Xie
 * @Date:   2020-01-16 12:54:54
 * @Last Modified by:   Haozhe Xie
 * @Last Modified time: 2020-01-16 12:55:01
 */

use Phalcon\Mvc\Model;

/**
 * The product in the application.
 *
 * @package TestZilla\model\Model\Product
 * @author Haozhe Xie <cshzxie@gmail.com>
 */
class Product extends Model {
    /**
     * Initialize the model.
     * According to the document of Phalcon framework, we have to announce 
     * the N-1 relationship in this function.
     */
    public function initialize() {
        $this->setSource(self::TABLE_NAME);
        $this->belongsTo('product_category_id', 'ProductCategory', 'product_category_id');
        $this->belongsTo('product_developer_id', 'User', 'uid');
        
        $this->hasMany('product_id', 'Issue', 'product_id');
    }

    /**
     * The getter of the field product_id.
     * @return the unique ID of the product
     */
    public function getProductId() {
        return $this->product_id;
    }

    /**
     * The setter of the field product_id.
     * @param long $productId - the unique ID of the product
     */
    public function setProductId($productId) {
        $this->product_id = $productId;
    }

    /**
     * The getter of the field product_name.
     * @return the name of the product in JSON format
     */
    public function getProductName() {
        return $this->product_name;
    }
    
    /**
     * The setter of the field product_name.
     * @param String $productName - the name of the product in JSON format
     */
    public function setProductName($productName) {
        $json = json_encode($productName);
        if ( json_last_error() != JSON_ERROR_NONE ) {
            throw new InvalidArgumentException('[Model\Product] The productName seems not a valid JSON.');
        }
        $this->product_name = $productName;
    }

    /**
     * The getter of the field product_logo.
     * @return the URL of the product logo
     */
    public function getProductLogo() {
        return $this->product_logo;
    }

    /**
     * The setter of the field product_logo.
     * @param String $productLogo - the URL of the product logo
     */
    public function setProductLogo($productLogo) {
        if ( mb_strlen($productLogo, 'utf-8') > 128 ) {
            throw new InvalidArgumentException('[Model\Product] The length of productLogo CANNOT exceed 128 characters.');
        }
        $this->product_logo = $productLogo;
    }

    /**
     * The getter of the field product_category_id.
     * @return the category of the product
     */
    public function getProductCategory() {
        return $this->productCategory;
    }

    /**
     * The setter of the field product_category_id.
     * @param ProductCategory $productCategry - the category of the product
     */
    public function setProductCategory(ProductCategory $productCategry) {
        if ( $productCategry == NULL ) {
            throw new InvalidArgumentException('[Model\Product] Invalid Model\ProductCategory.');
        }
        $this->product_category_id = $productCategry->getProductCategoryId();
    }

    /**
     * The getter of the field product_latest_version.
     * @return the latest version of the product
     */
    public function getLatestVersion() {
        return $this->product_latest_version;
    }

    /**
     * The setter of the field product_latest_version.
     * @param $latestVersion - the latest version of the product
     */
    public function setLatestVersion($latestVersion) {
        if ( mb_strlen($latestVersion, 'utf-8') > 24 ) {
            throw new InvalidArgumentException('[Model\Product] The length of latestVersion CANNOT exceed 24 characters.');
        }
        $this->product_latest_version = $latestVersion;
    }

    /**
     * The getter of the field product_developer_id.
     * @return the developer of the product
     */
    public function getDeveloper() {
        return $this->user;
    }

    /**
     * The setter of the field product_developer_id.
     * @param User $developer - the developer of the product
     */
    public function setDeveloper(User $developer) {
        if ( $developer == NULL ) {
            throw new InvalidArgumentException('[Model\Product] Invalid developer (Model\User).');
        }
        $this->product_developer_id = $developer->getUid();
    }

    /**
     * The getter of the field product_prerequisites.
     * @return the prerequisites of testing the product in JSON format
     */
    public function getPrerequisites() {
        return $this->product_prerequisites;
    }

    /**
     * The getter of the field product_prerequisites.
     * @param String $prerequisites - the prerequisites of testing the product in JSON format
     */
    public function setPrerequisites($prerequisites) {
        $json = json_encode($prerequisites);
        if ( json_last_error() != JSON_ERROR_NONE ) {
            throw new InvalidArgumentException('[Model\Product] The prerequisites seems not a valid JSON.');
        }
        $this->product_prerequisites = $prerequisites;
    }

    /**
     * The getter of the field product_url.
     * @return the URL to get the product
     */
    public function getProductUrl() {
        return $this->product_url;
    }

    /**
     * The setter of the field product_url.
     * @param String $productUrl - the URL to get the product
     */
    public function setProductUrl($productUrl) {
        if ( mb_strlen($productUrl, 'utf-8') > 256 ) {
            throw new InvalidArgumentException('[Model\Product] The length of productUrl CANNOT exceed 256 characters.');
        }
        $this->product_url = $productUrl;
    }

    /**
     * The getter of the field product_description.
     * @return the description of the product in JSON format
     */
    public function getDescription() {
        return $this->product_description;
    }

    /**
     * The setter of the field product_description.
     * @param String $description - the description of the product in JSON format
     */
    public function setDescription($description) {
        $json = json_encode($description);
        if ( json_last_error() != JSON_ERROR_NONE ) {
            throw new InvalidArgumentException('[Model\Product] The description seems not a valid JSON.');
        }
        $this->product_description = $description;
    }

    /**
     * Get number of issues of the product.
     * @return number of issues of the product
     */
    public function getNumberOfIssues() {
        return $this->issue->count();
    }

    /**
     * Get the description of the User object.
     * @return the description of the User object
     */
    public function __toString() {
        return sprintf('Product: [ID=%d, Name=%s, LogoUrl=%s, CategoryId=%d, LatestVersion=%s, DeveloperId=%d, URL=%s, Description=%s]', 
                $this->product_id, $this->product_name, $this->product_logo, $this->product_category_id, 
                $this->product_latest_version, $this->product_developer_id, $this->product_url, $this->product_description);
    }

    /**
     * The unique ID of the product.
     * @var long
     */
    protected $product_id;

    /**
     * The name of the product (JSON format).
     * @var String
     */
    protected $product_name;
    
    /**
     * The URL of the product logo.
     * @var String
     */
    protected $product_logo;
    
    /**
     * The unique ID of the product category.
     * @var int
     */
    protected $product_category_id;
    
    /**
     * The latest version of the product.
     * @var String
     */
    protected $product_latest_version;
    
    /**
     * The developer(User) of the product.
     * @var long
     */
    protected $product_developer_id;
    
    /**
     * The prerequisites of testing the product (JSON format).
     * @var String
     */
    protected $product_prerequisites;
    
    /**
     * The URL to get the product.
     * @var String
     */
    protected $product_url;
    
    /**
     * The description of the product (JSON format).
     * @var String
     */
    protected $product_description;

    /**
     * The table name of the model.
     */
    const TABLE_NAME = 'tz_products';
}